File: mobilev2/ios/Runner/NativeAdFactorySmall.swift
import Flutter
import GoogleMobileAds
import UIKit
import google_mobile_ads

/// Native ad factory SMALL (~80pt height) — banner-like.
/// XIB: `NativeAdSmallView.xib`
class NativeAdFactorySmall: FLTNativeAdFactory {

    func createNativeAd(
        _ nativeAd: GADNativeAd,
        customOptions: [AnyHashable: Any]? = nil
    ) -> GADNativeAdView? {
        guard let nibObjects = Bundle.main.loadNibNamed(
            "NativeAdSmallView", owner: nil, options: nil),
              let adView = nibObjects.first as? GADNativeAdView
        else {
            return nil
        }

        (adView.headlineView as? UILabel)?.text = nativeAd.headline

        if let body = nativeAd.body {
            (adView.bodyView as? UILabel)?.text = body
            adView.bodyView?.isHidden = false
        } else {
            adView.bodyView?.isHidden = true
        }

        if let cta = nativeAd.callToAction {
            (adView.callToActionView as? UIButton)?.setTitle(cta, for: .normal)
            adView.callToActionView?.isHidden = false
        } else {
            adView.callToActionView?.isHidden = true
        }

        if let icon = nativeAd.icon?.image {
            (adView.iconView as? UIImageView)?.image = icon
            adView.iconView?.isHidden = false
        } else {
            adView.iconView?.isHidden = true
        }

        adView.callToActionView?.isUserInteractionEnabled = false
        adView.nativeAd = nativeAd

        return adView
    }
}

File: mobilev2/android/app/src/main/kotlin/com/example/flutter_base2/NativeAdFactorySmall.kt
package com.example.flutter_base2

import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin

/**
 * Native ad factory SMALL (~80dp height) — banner-like.
 * Dùng khi `NativeAd(factoryId: 'nativeSmall', ...)` từ Dart.
 *
 * Layout: `res/layout/native_ad_small.xml`
 */
class NativeAdFactorySmall(
    private val layoutInflater: LayoutInflater
) : GoogleMobileAdsPlugin.NativeAdFactory {

    override fun createNativeAd(
        nativeAd: NativeAd,
        customOptions: MutableMap<String, Any>?
    ): NativeAdView {
        val view = layoutInflater.inflate(R.layout.native_ad_small, null) as NativeAdView

        val headline = view.findViewById<TextView>(R.id.ad_headline)
        val body = view.findViewById<TextView>(R.id.ad_body)
        val cta = view.findViewById<Button>(R.id.ad_call_to_action)
        val icon = view.findViewById<ImageView>(R.id.ad_icon)

        headline.text = nativeAd.headline
        view.headlineView = headline

        nativeAd.body?.let {
            body.text = it
            body.visibility = View.VISIBLE
        } ?: run { body.visibility = View.GONE }
        view.bodyView = body

        nativeAd.callToAction?.let {
            cta.text = it
            cta.visibility = View.VISIBLE
        } ?: run { cta.visibility = View.GONE }
        view.callToActionView = cta

        nativeAd.icon?.let {
            icon.setImageDrawable(it.drawable)
            icon.visibility = View.VISIBLE
        } ?: run { icon.visibility = View.GONE }
        view.iconView = icon

        view.setNativeAd(nativeAd)
        return view
    }
}
